llvm/include/llvm/Analysis/DXILMetadataAnalysis.h

```diff
@@ -19,12 +21,23 @@ namespace llvm {
 
 namespace dxil {
 
+struct EntryProperties {
+  const Function *Entry;
+  // Specific target shader stage may be specified for entry functions
+  Triple::EnvironmentType ShaderStage = Triple::UnknownEnvironment;
+  unsigned NumThreadsX{0}; // X component
+  unsigned NumThreadsY{0}; // Y component
+  unsigned NumThreadsZ{0}; // Z component
+
+  EntryProperties(const Function &Fn) : Entry(&Fn){};
+};
+
 struct ModuleMetadataInfo {
   VersionTuple DXILVersion{};
   VersionTuple ShaderModelVersion{};
   Triple::EnvironmentType ShaderStage = Triple::UnknownEnvironment;
   VersionTuple ValidatorVersion{};
-
+  SmallVector<EntryProperties, 4> EntryPropertyVec{};
```

Comment:
Unless there's a specific argument for 4, it's better to use `SmallVector<EntryProperties>` which will automatically deduce a reasonable size